core/services/workflows/engine.go

```diff
@@ -631,10 +622,21 @@ func (e *Engine) finishExecution(ctx context.Context, executionID string, status
 	return nil
 }
 
-func (e *Engine) workerForStepRequest(ctx context.Context, msg stepRequest) {
-	defer func() { e.newWorkerCh <- struct{}{} }()
+func (e *Engine) worker(ctx context.Context) {
+	e.wg.Add(1)
```

Comment:
@nolag It's more idiomatic to add outside of the goroutine; you can also do it all in one go, eg. e.wg.Add(100)

Reply:
This way is actually a race because `sync.Wait()` could be executed before this goroutine is scheduled and does the add.